File: 121.best-time-to-buy-and-sell-stock.py
#
# @lc app=leetcode id=121 lang=python
#
# [121] Best Time to Buy and Sell Stock
#

# @lc code=start
class Solution(object):
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        if len(prices) == 0:
            return 0
        soFarMin = prices[0]
        maxP = 0
        for i in range(1, len(prices)):
            if soFarMin > prices[i]:
                soFarMin = prices[i]
            else:
                maxP = max(maxP, prices[i] - soFarMin)
        return maxP
# @lc code=end

File: 121.best-time-to-buy-and-sell-stock.cpp
/*
 * @lc app=leetcode id=121 lang=cpp
 *
 * [121] Best Time to Buy and Sell Stock
 */
#include <vector>
using namespace std;

// @lc code=start
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int n = prices.size();
        if (n==0)
            return 0;
        int soFarMin = prices[0];
        int maxBenefit = 0;
        for (int i = 0; i < n; i++) {
            if (soFarMin > prices[i])
                soFarMin = prices[i];
            else
                maxBenefit = max(maxBenefit, prices[i] - soFarMin);
        }
        return maxBenefit;
    }
};
// @lc code=end

